v2/types/types.go

```diff
@@ -364,6 +365,50 @@ type Type struct {
 
 	// The underlying Go type.
 	GoType gotypes.Type
+
+	// The reference to Multiverse
+	multiverse *multiverse
+}
+
+// multiverse holds Type definitions that were not found in the imported code but are needed during
+// generation. For example the imported code my have T but not *T, while the generated code needs *T. This
+// can't be part of the main Universe because that is a simple map, with no locking, and we all know what happens
+// when you modify a map while it is being iterated. Storing the multiverse alongside a Type ensures that there's
+// at most one *Type for every type, and maintains the invariant that PointerTo(String) == PointerTo(String).
+type multiverse struct {
+	real      Universe
+	mu        sync.Mutex
+	synthetic map[string]*Type
+}
+
+// InitMultiverse inits a multiverse for a Type.
+// It panics if called twice.
+func (t *Type) InitMultiverse(u Universe) {
```

Comment:
Just call it Init -- multiverse is an implementation detail, it doesn't need to leak into the API